shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,776 @@
+          exists(Declaration target |
+            adjustedAccessType(a, apos, target, _, _) and
+            target.getDeclaredType(_, _) instanceof TypeParameter
+          )
+        }
+
+        pragma[nomagic]
+        private Type inferRootType(Access a, AccessPosition apos) {
+          relevantAccess(a, apos) and
+          result = a.getInferredType(apos, TypePath::nil())
+        }
+
+        pragma[nomagic]
+        private Type inferTypeAt(Access a, AccessPosition apos, TypeParameter tp, TypePath suffix) {
+          relevantAccess(a, apos) and
+          exists(TypePath path0 |
+            result = a.getInferredType(apos, path0) and
+            path0.isCons(tp, suffix)
+          )
+        }
+
+        /**
+         * Holds if `baseMention` is a (transitive) base type mention of the type of
+         * `a` at position `apos`, and `t` is mentioned (implicitly) at `path` inside
+         * `base`. For example, in
+         *
+         * ```csharp
+         * class C<T1> { }
+         *
+         * class Base<T2> { }
+         *
+         * class Mid<T3> : Base<C<T3>> { }
+         *
+         * class Sub<T4> : Mid<C<T4>> { }
+         *
+         * new Sub<int>().ToString();
+         * ```
+         *
+         * for the node `new Sub<int>()`, which is the receiver of a method call, we
+         * have:
+         *
+         * `baseMention` | `path`    | `t`
+         * ------------- | --------- | ---
+         * `Mid<C<T4>>`  | `"0"`     | ``C`1``
+         * `Mid<C<T4>>`  | `"0.1"`   | `int`
+         * `Base<C<T3>>` | `"0"`     | ``C`1``
+         * `Base<C<T3>>` | `"0.0"`   | ``C`1``
+         * `Base<C<T3>>` | `"0.0.1"` | `int`
+         */
+        pragma[nomagic]
+        predicate hasBaseTypeMention(
+          Access a, AccessPosition apos, TypeMention baseMention, TypePath path, Type t
+        ) {
+          exists(Type sub | sub = inferRootType(a, apos) |
+            baseTypeMentionHasNonTypeParameterAt(sub, baseMention, path, t)
+            or
+            exists(TypePath prefix, TypePath suffix, TypeParameter tp |
+              baseTypeMentionHasTypeParameterAt(sub, baseMention, prefix, tp) and
+              t = inferTypeAt(a, apos, tp, suffix) and
+              path = prefix.append(suffix)
+            )
+          )
+        }
+      }
+
+      pragma[nomagic]
+      private predicate accessBaseType(
+        Access a, AccessPosition apos, Declaration target, Type base, TypePath path, Type t
+      ) {
+        exists(TypeMention tm |
+          target = a.getTarget() and
+          AccessBaseType::hasBaseTypeMention(a, apos, tm, path, t) and
+          base = resolveTypeMentionRoot(tm)
+        )
+      }
+
+      pragma[nomagic]
+      private predicate declarationBaseType(
+        Declaration decl, DeclarationPosition dpos, Type base, TypePath path, Type t
+      ) {
+        t = decl.getDeclaredType(dpos, path) and
+        path.isCons(base.getATypeParameter(), _)
```

Comment:
This relies crucially on `TypeParameter`s being unique to a given `Type`. Should that be checked in a consistency check somewhere?

Reply:
The rewrite will not work if a declaration has multiple declared types at `dpos`, which there is no reason to rule out. But I noticed that `declarationBaseType` is only used in a context where `t` is in fact a `TypeParameter`, so we may as well specialize it.

Reply:
> This relies crucially on TypeParameters being unique to a given Type. Should that be checked in a consistency check somewhere?

Good idea; not all type parameters need to belong to a type though (e.g. method type parameters in C#), but they should belong to at most one type.